Feat/provide form model

marina-mosti opened the pull request:
Refactor FVL to use a provide/inject model instead of `v-model` to prevent the whole form from re-rendering when a piece of the model changes

marina-mosti committed:
feat: create a SchemaField component

marina-mosti committed:
feat: move formModel updating to provide inject

In order to avoid full form refreshing on each input update, we moved onto provide inject with an external model

marina-mosti committed:
feat: update form factory tests for new component merge

marina-mosti committed:
update: SchemaWizard to use new useSchemaForm composable

tzhelyazkova commented on src/features/useSchemaForm.js:
```diff
@@ -0,0 +1,40 @@
+import { ref, unref, provide } from 'vue'
+
+export default function useSchemaForm (initialFormValue = {}) {
+  const formModel = ref(unref(initialFormValue))
```
hmm what is this doing?

marina-mosti replied:
The original intent was to make sure that the initial form value was a ref, but ill refactor with `isRef` to check and not break the initial reactivity

marina-mosti committed:
refactor: schema injection into sub components

The schema was being passed down, but it presented quite a few bugs when introducing a computed schema in a deeply nested example.
Additionally, v-model modelValue injection into subcomponents from SchemaField was completely broken.

marina-mosti committed:
3.0.0-beta.5

marina-mosti committed:
fix: add missing useSchemaForm export

marina-mosti committed:
3.0.0-beta.6

marina-mosti committed:
refactor: reactivity check for initialFormValue

marina-mosti committed:
update: vue and vtu versions

marina-mosti committed:
fix: add emits for demo comps

marina-mosti committed:
feat: add SchemaField units and remove div wrap

marina-mosti committed:
fix: SchemaWizard different provide string

SchemaWizard was hijacking the new logic that allows for SchemaForm to determine how the schema ref needs to be provided to sub schemas, so a new string should suffice to let nested SchemaForms know that they are part of a wizard in order to correctly display as a div and conceal the slots.

marina-mosti committed:
v3.0.0-beta.7

marina-mosti committed:
fix: wiz should prevent model and add missing name

marina-mosti committed:
Feat/v3 docs (#151)

* feat: create 3.x docs

* feat: move 2.x docs to folder

* fix: wizard should preserve schema

* feat: export useSchemaForm

* feat: update main section of 3.x docs

* feat: update plugin docs

* feat: add a11y video to docs

* feat: add version menus to 2x and 3x docs

* feat: update examples with 3.x

* feat: add nested generated example

* update: 100 nested computed example

* feat: add migration guide from 2.x (#155)

shentao commented on src/SchemaForm.vue:
```diff
@@ -59,19 +56,43 @@ export default {
+    const behaveLikeParentSchema = computed(() => (!isChildOfWizard && !hasParentSchema))
+
+    const { schema } = toRefs(props)
+    let injectedSchema = inject('injectedSchema', false)
```
**suggestion:** Given that this is an OSS lib it might be good to use a more unique string or symbol for provide/inject to avoid accidental conflicts done by users that might not be aware of the internal workings of this library. Maybe something like a `fvl_` prefix?

shentao reviewed (commented):
Looks solid! Nice work! One thing that might be nice to explore is how we could improve the plugin architecture now that `SchemaField` component is available. Maybe this could simplify some of the code inside Vuelidate plugin? Since we no longer need yet another wrapper component around the form field, since there’s already a wrapper that could handle that and more. Maybe the plugins could inject some composables that `SchemaField` would execute and return? Just throwing ideas as I don’t have any specific idea how to implement that.
